src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -192,27 +300,76 @@ static ucs_status_t uct_rdmacm_cm_create_dummy_qp(struct rdma_cm_id *id,
     return UCS_OK;
 }
 
+static ucs_status_t uct_rdamcm_cm_ep_create_qpn(uct_rdmacm_cm_ep_t *cep,
+                                                struct uct_rdmacm_cm_qp_context *ctx)
+{
+    ucs_status_t status;
+
+    if (ctx->reserved_qpn_support) {
+        status = uct_rdamcm_cm_ep_create_reserved_qpn(cep, ctx);      
+    } else {
+        /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
+        status = uct_rdmacm_cm_create_dummy_qp(cep->id, ctx->dummy_qp_ctx.cq,
+                                               &cep->qp);
+    }
+
+    if (status == UCS_OK) {
+        cep->flags |= UCT_RDMACM_CM_EP_QPN_CREATED;
```

Comment:
why need this flag?

Reply:
@yosefe As we talked in a previous email, the cleanup function can be called at very beginning before ibverbs initilized. As you suggested that we should only destroy qpn after it was created. So add this flag to indicate qpn was created or not.

Reply:
but if we create cep successfully, it means uct_rdamcm_cm_ep_set_qp_num() returned OK, so we must have created QPN successfully too. So how we can have a 'cep' without a QPN?

Reply:
@yosefe cep is a parameter of uct_rdamcm_cm_ep_set_qp_num(), so it existed before calling uct_rdamcm_cm_ep_set_qp_num(). In uct_rdamcm_cm_ep_set_qp_num we just get a QPN according to the info in cep parameter(such as cep->id/cep->verbs, etc) and store to cep->qpn and then assign it to conn_param->qp_num. cep exists from UCS_CLASS_INIT_FUNC, we have the code in UCS_CLASS_INIT_FUNC to init the member values of cep.
Maybe I misunderstood your question?

Reply:
ok.
can we write it like this
```
if (status != UCS_OK) {
    return status;
}

cep->flags |= UCT_RDMACM_CM_EP_QPN_CREATED;
```

Reply:
@yosefe Fixed.